pkg/envoy/ads/response.go

```diff
@@ -42,45 +43,50 @@ func (s *Server) sendTypeResponse(tURI envoy.TypeURI,
 	return nil
 }
 
-func (s *Server) sendAllResponses(proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, cfg configurator.Configurator) {
-	log.Trace().Msgf("A change announcement triggered *DS update for proxy with SerialNumber=%s on Pod with UID=%s", proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-
-	// Tracks the success of this full update of all its XDS paths. If a single XDS response path fails for this full update,
-	// the full updated will be considered as failed for metric purposes (success = false)
+// sendResponse takes a set of TypeURIs which will be called to generate the xDS resources
+// for, and will have them sent to the proxy server.
+// If no DiscoveryRequest is passed, an empty one for the TypeURI is created
+func (s *Server) sendResponse(typeURIsToSend mapset.Set,
+	proxy *envoy.Proxy,
+	server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
+	request *xds_discovery.DiscoveryRequest,
+	cfg configurator.Configurator) error {
 	success := true
-	defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	if typeURIsToSend.Cardinality() == len(envoy.XDSResponseOrder) {
+		defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	}
 
 	// Order is important: CDS, EDS, LDS, RDS
 	// See: https://github.com/envoyproxy/go-control-plane/issues/59
 	for _, typeURI := range envoy.XDSResponseOrder {
-		// For SDS we need to add ResourceNames
-		var request *xds_discovery.DiscoveryRequest
-		if typeURI == envoy.TypeSDS {
-			request = makeRequestForAllSecrets(proxy, s.catalog)
-			if request == nil {
-				continue
+		if !typeURIsToSend.Contains(typeURI) {
```

Comment:
Instead of iterating over a slice, then checking for membership of a set - can we not pass the precise slice we want?
This would be much simpler if instead of having both `envoy.XDSResponseOrder` and `typeURIsToSend` - we had only one ordered slice, which you can pass into `sendResponse()`.

Reply:
+1, just `[]string{..}` doesn't suffice?

Reply:
This is made so to have xDS ordering depend on the input slice. I accounted that if we walk the input slice, this could be prone to error and ordering will depend on the callee.

If that's fine by you however, fine by me.

Reply:
I took a stab at this here:  https://github.com/openservicemesh/osm/pull/3156